contracts/staking/PoSValidatorManager.sol

```diff
@@ -136,26 +142,53 @@ abstract contract PoSValidatorManager is
         uint32 messageIndex
     ) external {
         PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
+        Validator memory validator = _initializeEndValidation(validationID);
+
         // Check that minimum stake duration has passed
-        Validator memory validator = getValidator(validationID);
         require(
-            block.timestamp
+            validator.endedAt
                 >= validator.startedAt + $._validatorRequirements[validationID].minStakeDuration,
             "PoSValidatorManager: minimum stake duration not met"
         );
 
         if (includeUptimeProof) {
-            _getUptime(validationID, messageIndex);
+            // Uptime proofs include the absolute number of seconds the validator has been active.
+            uint64 uptimeSeconds = _getUptime(validationID, messageIndex);
+            // Save this value for use by this validator's delegators.
+            $._completedValidationUptimeSeconds[validationID] = uptimeSeconds;
+
+            $._redeemableValidatorRewards[validationID] += $._rewardCalculator.calculateReward({
+                stakeAmount: weightToValue(validator.startingWeight),
+                validatorStartTime: validator.startedAt,
+                stakingStartTime: validator.startedAt,
+                stakingEndTime: validator.endedAt,
+                uptimeSeconds: uptimeSeconds,
+                initialSupply: 0,
+                endSupply: 0
+            });
         }
-        // TODO: Calculate the reward for the validator, but do not unlock it
-
-        _initializeEndValidation(validationID);
     }
 
     function completeEndValidation(uint32 messageIndex) external {
         PoSValidatorManagerStorage storage $ = _getPoSValidatorManagerStorage();
+
         (bytes32 validationID, Validator memory validator) = _completeEndValidation(messageIndex);
-        _unlock(validator.startingWeight, $._validatorRequirements[validationID].owner);
+
+        // Return now if this was originally a PoA validator that was later migrated to this PoS manager
+        if (!_isPoSValidator(validationID)) {
+            return;
+        }
+
+        address owner = $._validatorRequirements[validationID].owner;
+        if (validator.status == ValidatorStatus.Completed) {
```

Comment:
In what case here would `validator.status` not be `Completed`? Could we add it as a comment

Reply:
Ah, I see below it's if the validator wasn't properly registered on the P-Chain prior to the message `expiry`. I'm wondering if we could remove the `Invalidated` status and just go off the `startTime` being zero in the contract, but not a priority right now.